File: skylib-c++/src-gtk4-pilot/gtk_gui_pilot.cpp
/*
 * gtk_pilot.cpp
 *
 *  Created on: Feb 26, 2021
 *      Author: scaille
 */

#include <utils.hh>
#include "gtk_entry_pilot.hh"
#include "gtk_gui_pilot.hh"

namespace ch_skymarshall::gui::gtk4::pilot {

gtk_gui_pilot::gtk_gui_pilot(Gtk::Window *_window) :
		m_window(_window) {

}

shared_ptr<gtk_entry_pilot> gtk_gui_pilot::entry(const string &_name) {
	return make_shared<gtk_entry_pilot>(this, _name);
}

Gtk::Widget* gtk_gui_pilot::find_widget(const string &_name) {
	Gtk::Widget *found = find(m_window->get_child(), _name);
	if (found == NULL) {
		throw gui_pilot_exception("No widget found: " + _name);
	}
	return found;
}

Gtk::Widget* gtk_gui_pilot::find(Gtk::Widget *_widget, const ustring &_name) {
	if (_name == _widget->get_name()) {
		return _widget;
	}
	if (_widget->get_first_child() == NULL) {
		return NULL;
	}
	Gtk::Widget *found = NULL;
	for (Gtk::Widget *child = _widget->get_first_child(); found == NULL && child != NULL; child =
			child->get_next_sibling()) {
		found = find(child, _name);
	}
	return found;
}

gtk_gui_pilot::~gtk_gui_pilot() = default;
}

File: skylib-rust/src/gtk_pilot/gtk_gui_pilot.rs
//! gtk_gui_pilot.rs
//!
//! Rust port of `gtk_gui_pilot.hh` + `.cpp` (originally `gtk_pilot.h`/`.cpp`).
//!
//! Original author: scaille (Feb 26, 2021)

use std::rc::Rc;

use gtk4::prelude::*;

use super::gtk_entry_pilot::GtkEntryPilot;
use crate::lib_properties::GuiException;

/// Port of `gui_pilot_exception`. The original added nothing over its base
/// beyond a `DESTR_LOG`'d destructor, so this is just a thin constructor
/// wrapping [`GuiException`] rather than a distinct type.
pub fn gui_pilot_exception(message: impl Into<String>) -> GuiException {
    GuiException::new(message)
}

/// Port of `gtk_gui_pilot`.
pub struct GtkGuiPilot {
    window: gtk4::Window,
}

impl GtkGuiPilot {
    pub fn new(window: gtk4::Window) -> Rc<Self> {
        Rc::new(Self { window })
    }

    /// Port of `gtk_gui_pilot::entry`.
    pub fn entry(self: &Rc<Self>, name: impl Into<String>) -> Rc<GtkEntryPilot> {
        GtkEntryPilot::new(Rc::clone(self), name.into())
    }

    /// Port of `gtk_gui_pilot::find_widget`.
    pub fn find_widget(&self, name: &str) -> Result<gtk4::Widget, GuiException> {
        let child = self
            .window
            .child()
            .ok_or_else(|| gui_pilot_exception(format!("No widget found: {name}")))?;
        Self::find(&child, name).ok_or_else(|| gui_pilot_exception(format!("No widget found: {name}")))
    }

    /// Port of the private recursive `gtk_gui_pilot::find`.
    fn find(widget: &gtk4::Widget, name: &str) -> Option<gtk4::Widget> {
        if widget.widget_name().as_str() == name {
            return Some(widget.clone());
        }
        let mut child = widget.first_child();
        while let Some(c) = child {
            if let Some(found) = Self::find(&c, name) {
                return Some(found);
            }
            child = c.next_sibling();
        }
        None
    }
}
